sched/wdog/wd_start.c

```diff
@@ -214,132 +276,104 @@ int wd_start(FAR struct wdog_s *wdog, sclock_t delay,
+      /* Resume the interval timer that will generate the next
+       * interval event. If the timer at the head of the list changed,
+       * then this will pick that new delay.
        */
 
-      else
-        {
-          delay -= now;
-          if (!curr->next)
-            {
-              sq_addlast((FAR sq_entry_t *)wdog, &g_wdactivelist);
-            }
-          else
-            {
-              next = curr->next;
-              next->lag -= delay;
-              sq_addafter((FAR sq_entry_t *)curr, (FAR sq_entry_t *)wdog,
-                          &g_wdactivelist);
-            }
-        }
+      nxsched_reassess_timer();
     }
+#else
+  UNUSED(reassess);
 
-  /* Put the lag into the watchdog structure and mark it as active. */
-
-  wdog->lag = delay;
+  /* Check if the watchdog has been started. If so, delete it. */
 
-#ifdef CONFIG_SCHED_TICKLESS
-  /* Resume the interval timer that will generate the next interval event.
-   * If the timer at the head of the list changed, then this will pick that
-   * new delay.
-   */
+  if (WDOG_ISACTIVE(wdog))
+    {
+      list_delete(&wdog->node);
+      wdog->func = NULL;
+    }
 
-  nxsched_resume_timer();
+  wd_insert(wdog, ticks, wdentry, arg);
 #endif
-
   leave_critical_section(flags);
   return OK;
 }
 
+/****************************************************************************
+ * Name: wd_start
+ *
+ * Description:
+ *   This function adds a watchdog timer to the active timer queue.  The
+ *   specified watchdog function at 'wdentry' will be called from the
+ *   interrupt level after the specified number of ticks has elapsed.
+ *   Watchdog timers may be started from the interrupt level.
+ *
+ *   Watchdog timers execute in the address environment that was in effect
+ *   when wd_start() is called.
+ *
+ *   Watchdog timers execute only once.
+ *
+ *   To replace either the timeout delay or the function to be executed,
+ *   call wd_start again with the same wdog; only the most recent wdStart()
+ *   on a given watchdog ID has any effect.
+ *
+ * Input Parameters:
+ *   wdog     - Watchdog ID
+ *   delay    - Delay count in clock ticks
+ *   wdentry  - Function to call on timeout
+ *   arg      - Parameter to pass to wdentry
+ *
+ *   NOTE:  The parameter must be of type wdparm_t.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success; a negated errno value is return to
+ *   indicate the nature of any failure.
+ *
+ * Assumptions:
+ *   The watchdog routine runs in the context of the timer interrupt handler
+ *   and is subject to all ISR restrictions.
+ *
+ ****************************************************************************/
+
+int wd_start(FAR struct wdog_s *wdog, sclock_t delay,
+             wdentry_t wdentry, wdparm_t arg)
+{
+  /* Verify the wdog and setup parameters */
+
+  if (delay < 0)
+    {
+      return -EINVAL;
+    }
+
+  return wd_start_absolute(wdog, clock_systime_ticks() + delay,
```

Comment:
```suggestion
  return wd_start_absolute(wdog, clock_systime_ticks() + delay,
```
I have a question here.
`clock_systime_ticks() + delay` should be executed in the critical section. If ISR triggered after the absolute tick increases, the delay time may expire early.